### PR TITLE
Stop propt inheriting from messaget [blocks: #3800]

### PR DESCRIPTION
**Pull request by romainbrenguier**
No functional changes.
A propt object should not be used as a messaget so the messaget should
rather be a field.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [na] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/solvers/prop/prop_conv.h [hunk 1]
```diff
@@ -117,7 +117,6 @@ class prop_conv_solvert:public prop_convt
   typedef std::map<irep_idt, literalt> symbolst;
   typedef std::unordered_map<exprt, literalt, irep_hash> cachet;
 
-  const cachet &get_cache() const { return cache; }
```
Comment (tautschnig) [inline src/solvers/prop/prop_conv.h]: Unrelated change?

**Review by tautschnig — CHANGES_REQUESTED** [review]
Overall I think this is a very good idea.
1) The reason for me requesting changes is the (seemingly?) unrelated change as noted below.
2) More of a comment than a strict requirement: let's not do this `set_message_handler` game, but instead make the `propt` constructor take a `message_handlert&` and set things up in the constructor.

**Comment by kroening** [comment]
This will introduce a lot of noise, if done consistently, and the benefit isn't obvious?

**Comment by tautschnig** [comment]
To me, the benefit would be fixing a design flaw (the "is-a" relationship seems wrong), but maybe I'm missing something?

**Review by peterschrammel — COMMENTED** [review]
I'm generally in favour of doing that if we do it across the entire code base. It's a lot of churn, but not really much breaking changes as we are mostly passing around message_handlert.
We should then also add a rule to the coding standard that (ui_)message_handlert is passed through the constructor instead of a setter and decide on a consistent name for the messaget member, eg one of `message`, `msg`, `log` or `logger`.

**Comment by tautschnig** [comment]
+1 to all that @peterschrammel said, but also I'd like to hear more from @kroening as he seemed to be less happy with the route being taken.

File: src/solvers/prop/prop_conv.h (hunk 1 above)
Comment (romainbrenguier) [reply src/solvers/prop/prop_conv.h]: yes sorry, this shouldn't be in this PR

**Comment by romainbrenguier** [comment]
@peterschrammel I'm not sure why we have to do it across the whole code base at the same time? Is it just a naming issue? `message` is the natural name for an object of type `messaget` but maybe `messaget` is not well chosen, `loggert` would have been better.

**Comment by martin-cs** [comment]
Kind of agree with the discussion above.  Let's be consistent; either inherit from or parameter to construction (do we ever swap the message output source?  If not, let's just have it as part of the constructor), a slow migration seems painful.  Is it worth doing ... unclear to me.

**Comment by romainbrenguier** [comment]
@martin-cs 
> Kind of agree with the discussion above. Let's be consistent; either inherit from or parameter to construction (do we ever swap the message output source? If not, let's just have it as part of the constructor), a slow migration seems painful. Is it worth doing ... unclear to me.

For me it's clearly worth doing. If someone is trying to understand this code, the first thing they see is what the class inherits from, then they wonder "why is a decision procedure a message?", then they dig into messaget and realize it doesn't really represent messages but still has no link with what a solver should intuitively do, and they are left confused before having learned anything useful about solvers.

**Comment by tautschnig** [comment]
From me to change my review to "approve" I'd ask for a proper constructor to be used, the member really shouldn't (need to) be public. As far as naming is concerned, let's try to set a precedent here: whatever is chosen here should make people happy and should over time be used elsewhere. (Personally I think something along the lines of `logger` is better, but mostly I'd just like to have consistency over time, whatever the name.)

**Comment by romainbrenguier** [comment]
@tautschnig I've added fixup commits to make the field protected, rename it to `log` (we use `log` in `goto_symext` already), and set the message handler through the constructor where needed. It's maybe better not to review commit by commit now as they will be squashed into just one.
Another point we may want to discuss is whether `log` should be mutable or not.

File: src/solvers/prop/prop.h [hunk 2]
```diff
@@ -21,10 +21,11 @@ Author: Daniel Kroening, kroening@kroening.com
 
 /*! \brief TO_BE_DOCUMENTED
 */
-class propt:public messaget
+class propt
 {
 public:
   propt() { }
```
Comment (tautschnig) [inline src/solvers/prop/prop.h:27]: Hmm, don't you need to consume a `message_handlert` here?

File: src/solvers/sat/dimacs_cnf.cpp [hunk 3]
```diff
@@ -19,6 +19,11 @@ dimacs_cnft::dimacs_cnft():break_lines(false)
 {
 }
 
+dimacs_cnft::dimacs_cnft(message_handlert &message_handler) : dimacs_cnft()
+{
+  log.set_message_handler(message_handler);
```
Comment (tautschnig) [inline src/solvers/sat/dimacs_cnf.cpp:24]: This is going to be a bit painful, but let's please try to use the constructor all the way, i.e., it would be `dimacs_cnft(message_handler)` above.

File: src/solvers/sat/dimacs_cnf.h [hunk 4]
```diff
@@ -18,6 +18,7 @@ class dimacs_cnft:public cnf_clause_listt
 {
 public:
   dimacs_cnft();
```
Comment (tautschnig) [inline src/solvers/sat/dimacs_cnf.h:20]: Please remove this constructor.

**Review by tautschnig — COMMENTED** [review]
As stated below, I'd propose an even stricter regime and only keep constructors that pass the message handler.

**Review by allredj — COMMENTED** [review]
:warning:
Diffblue compatibility check is currently unavailable.
Please create manual bump.
(cbmc commit: b7aeac96e73bdc405fafaa2fc74c39e0e499abcf).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90838766

File: src/solvers/prop/prop.h (hunk 2 above)
Comment (romainbrenguier) [reply src/solvers/prop/prop.h:27]: I wasn't there before so adding one instead of constructing a new one could potentially change the behavior.

File: src/solvers/sat/dimacs_cnf.cpp (hunk 3 above)
Comment (romainbrenguier) [reply src/solvers/sat/dimacs_cnf.cpp:24]: That sounds like a good idea but this would go a bit beyond scope of a refactoring PR, and my initial statement "No functional changes." wouldn't be strictly true anymore.

File: src/solvers/prop/prop.h (hunk 2 above)
Comment (tautschnig) [reply src/solvers/prop/prop.h:27]: Yes, it makes things safer, so I'd happily take that change.

File: src/solvers/sat/dimacs_cnf.cpp (hunk 3 above)
Comment (tautschnig) [reply src/solvers/sat/dimacs_cnf.cpp:24]: Well, if we have no bugs (surely true) then it wouldn't entail any functional changes, because each time we construct a `dimacs_cnft` we surely also initialized its message handler.

Comment (romainbrenguier) [reply src/solvers/sat/dimacs_cnf.cpp:24]: It we leave it initialized, doesn't it just means that some messages wouldn't get printed? This would be difficult to detect. It's also a bit difficult to manually check, as a lot of classes inherit from `dimacs_cnft`...

Comment (tautschnig) [reply src/solvers/sat/dimacs_cnf.cpp:24]: I think any introduced functional change is a strict improvement here. Any class deriving should be forced to call a constructor that takes a message handler.

**Review by LAJW — APPROVED** [review]
Magnificent

**Review by tautschnig — APPROVED** [review]
Approving for now (but needs a rebase), the changes I had requested will happen as part of #3800.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: de1a8f05ad15e1172adf4f555500bf7406934cf4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99395288

File: src/solvers/sat/satcheck_minisat2.cpp [hunk 5]
```diff
@@ -238,21 +238,21 @@ propt::resultt satcheck_minisat2_baset<T>::prop_solve()
 
         if(solver_result==l_True)
         {
-          messaget::status() <<
-            "SAT checker: instance is SATISFIABLE" << eom;
+          log.status() << "SAT checker: instance is SATISFIABLE"
```
Comment (tautschnig) [inline src/solvers/sat/satcheck_minisat2.cpp:241]: Line 230 also needs to be updated (Windows build).

File: src/solvers/flattening/arrays.cpp [hunk 6]
```diff
@@ -580,7 +580,7 @@ void arrayst::add_array_constraints_update(
 
     if(index_expr.type()!=value.type())
     {
-      prop.error() << expr.pretty() << messaget::eom;
+      prop.message.error() << expr.pretty() << messaget::eom;
```
Comment (tautschnig) [inline src/solvers/flattening/arrays.cpp:583]: This change (inside a `#if 0` block) looks suspicious: I believe this should be just `error()`.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 756b0036d6d6fd3c8b9c3a6348f06ad498cd0431).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99448369
